Doc/library/typing.rst

```diff
@@ -933,6 +933,55 @@ These can be used as types in annotations using ``[]``, each having a unique syn
 
    .. versionadded:: 3.9
 
+
+.. data:: TypeGuard
+
+   Special typing form used to annotate the return type of a user-defined
+   type guard function.  ``TypeGuard`` only accepts a single type argument.
+
+   ``TypeGuard`` aims to benefit *type narrowing* - a technique used by static
+   type checkers to determine a more precise type of an expression within a
+   program's code flow.  Usually type narrowing is done by analyzing
+   conditional code flow and applying the narrowing to a block of code.  The
+   conditional expression here is sometimes referred to as a "type guard".
+
+   Sometimes, a type guard can be a complex checking function.
+   These type guard functions require ``TypeGuard`` to narrow their input types
+   as the static type checker usually does not have enough information to
+   statically infer them.
+
+   A ``TypeGuard`` tells the static type checker that for a given function:
+      1. The return value is a boolean.
+      2. If the return value was "truthy", the type of the input to the
+         function is specified by the type inside ``TypeGuard``.
+
+      For example::
+
+         def is_str_list(val: List[object]) -> TypeGuard[List[str]]:
+             '''Determines whether all objects in the list are strings'''
+             return all(isinstance(x, str) for x in val)
+
+         def func1(val: List[object]):
+             if is_str_list(val):
+                 # Type of ``val`` is narrowed to List[str]
+                 print(" ".join(val))
+             else:
+                 # Type of ``val`` remains as List[object]
+                 print("Not a list of strings!")
+
+   In short, the form ``def foo(arg: TypeA) -> TypeGuard[TypeB]: ...``,
+   Means that if ``foo(arg)`` returned true, then ``arg`` narrows from
+   ``TypeA`` to ``TypeB``.
```

Comment:
I'd add something explaining that this "narrowing" could in practice be a "widening" in case `TypeA` is a subclass of `TypeB`. (I think there's an example in the PEP, as well as an explanation why the checker should not make an exception in this case.)